File: mobile-apps/android/app/src/main/java/com/ahmadjalil/tcger/feature/libraryoperations/LibraryOperationsModels.kt
package com.ahmadjalil.tcger.feature.libraryoperations

import kotlinx.serialization.Serializable

@Serializable
data class StoragePlacement(
    val id: String,
    val collectionEntryId: String,
    val slotIndex: Int,
    val quantity: Int,
    val stackKey: String? = null,
    val cardName: String? = null,
    val printedName: String? = null,
)

@Serializable
data class StorageCompartment(
    val id: String,
    val label: String,
    val order: Int,
    val pageNumber: Int? = null,
    val rows: Int,
    val columns: Int,
    val capacity: Int,
    val locked: Boolean = false,
    val placements: List<StoragePlacement> = emptyList(),
)

@Serializable
data class StorageContainer(
    val id: String,
    val binderId: String? = null,
    val name: String,
    val kind: String,
    val order: Int,
    val isUnsorted: Boolean = false,
    val locked: Boolean = false,
    val compartments: List<StorageCompartment> = emptyList(),
)

@Serializable
data class CreateStorageContainerRequest(
    val name: String,
    val kind: String,
    val binderId: String? = null,
    val order: Int? = null,
    val isUnsorted: Boolean = false,
    val locked: Boolean = false,
)

@Serializable
data class UpdateStorageContainerRequest(
    val name: String? = null,
    val order: Int? = null,
    val locked: Boolean? = null,
)

@Serializable
data class CreateStorageCompartmentRequest(
    val containerId: String,
    val label: String,
    val order: Int,
    val pageNumber: Int? = null,
    val rows: Int,
    val columns: Int,
    val capacity: Int,
    val locked: Boolean = false,
)

@Serializable
data class UpdateStorageCompartmentRequest(
    val label: String? = null,
    val order: Int? = null,
    val pageNumber: Int? = null,
    val locked: Boolean? = null,
)

@Serializable
data class PlaceCollectionEntryRequest(
    val compartmentId: String,
    val collectionEntryId: String,
    val slotIndex: Int,
    val quantity: Int,
    val allowDuplicateStacking: Boolean = false,
)

/**
 * Fast client-side checks for the storage editor. The server remains authoritative
 * for ownership, owned quantity, and whether two entries represent the same printing.
 */
object StoragePlacementRules {
    fun placementError(
        container: StorageContainer,
        compartment: StorageCompartment,
        slotIndex: Int,
        quantity: Int,
        allowDuplicateStacking: Boolean,
        movingPlacementId: String? = null,
    ): String? {
        if (container.locked) return "Container is locked"
        if (compartment.locked) return "Compartment is locked"
        if (slotIndex !in 0 until compartment.capacity) {
            return "Slot must be between 1 and ${compartment.capacity}"
        }
        if (quantity < 1) return "Quantity must be at least 1"
        val occupants = compartment.placements.filter {
            it.slotIndex == slotIndex && it.id != movingPlacementId
        }
        if (occupants.isNotEmpty() && !allowDuplicateStacking) {
            return "Slot ${slotIndex + 1} is occupied; enable duplicate stacking to use it"
        }
        return null
    }

    fun compartmentError(rows: Int, columns: Int, capacity: Int): String? = when {
        rows < 1 -> "Rows must be at least 1"
        columns < 1 -> "Columns must be at least 1"
        capacity < 1 -> "Capacity must be at least 1"
        capacity > rows * columns -> "Capacity cannot exceed rows × columns"
        else -> null
    }
}

@Serializable
data class DeckCheckoutAllocation(
    val id: String,
    val deckCardId: String,
    val collectionEntryId: String,
    val quantity: Int,
    val containerId: String? = null,
    val containerName: String? = null,
    val compartmentId: String? = null,
    val compartmentLabel: String? = null,
    val slotIndex: Int? = null,
    val cardName: String? = null,
    val printedName: String? = null,
    val refilledAt: String? = null,
) {
    val locationDescription: String
        get() = buildList {
            containerName?.let(::add)
            compartmentLabel?.let(::add)
            slotIndex?.let { add("Slot ${it + 1}") }
        }.joinToString(" · ").ifBlank { "Unsorted" }
}

@Serializable
data class DeckCheckoutSession(
    val id: String,
    val deckId: String,
    val status: String,
    val note: String? = null,
    val checkedOutAt: String,
    val checkedInAt: String? = null,
    val allocations: List<DeckCheckoutAllocation> = emptyList(),
) {
    val isCheckedOut: Boolean get() = status == "checked_out"
}

@Serializable data class DeckCheckoutRequest(val note: String? = null)

@Serializable
data class RapidSetEntryRequest(
    val binderId: String,
    val tcg: String,
    val setCode: String,
    val entries: List<RapidSetEntryRow>,
)

@Serializable
data class RapidCardData(
    val name: String,
    val printedName: String? = null,
    val searchAliases: List<String>? = null,
    val tcg: String,
    val externalId: String,
    val setCode: String? = null,
    val setName: String? = null,
    val collectorNumber: String? = null,
    val imageUrl: String? = null,
    val imageUrlSmall: String? = null,
)

@Serializable
data class RapidSetEntryRow(
    val rowId: String,
    val collectorNumber: String,
    val card: RapidCardData,
    val quantity: Int = 1,
)

@Serializable
data class RapidSetEntryReceipt(
    val receiptId: String,
    val addedRows: Int,
    val addedCopies: Int,
    val items: List<RapidSetEntryItem>,
)

@Serializable
data class RapidSetEntryItem(
    val rowId: String,
    val collectorNumber: String,
    val entryId: String,
    val auditId: String,
    val quantity: Int,
)

@Serializable data class UndoCollectionMutationRequest(val idempotencyKey: String)

@Serializable data class RapidCardSearchResponse(val cards: List<RapidCardData> = emptyList(), val total: Int = cards.size)

@Serializable
data class AcquisitionCostSplitItem(
    val collectionEntryId: String,
    val weight: Int = 1,
)

@Serializable
data class AcquisitionCostSplitRequest(
    val totalCents: Long,
    val currency: String,
    val mode: String,
    val lines: List<AcquisitionCostSplitItem>,
    val notes: String? = null,
)

@Serializable
data class AcquisitionCostAllocation(
    val collectionEntryId: String,
    val allocatedCents: Long,
    val acquisitionPrice: Double,
    val transactionId: String,
)

@Serializable
data class AcquisitionCostSplitReceipt(
    val allocationGroupId: String,
    val auditId: String,
    val totalCents: Long,
    val currency: String,
    val allocations: List<AcquisitionCostAllocation>,
)

@Serializable
data class PsaCertificationLookup(
    val certNumber: String,
    val grader: String,
    val grade: Double? = null,
    val gradeLabel: String? = null,
    val labelType: String? = null,
    val year: String? = null,
    val brand: String? = null,
    val subject: String? = null,
    val searchableName: String? = null,
    val cardNumber: String? = null,
    val variety: String? = null,
    val category: String? = null,
    val population: Int? = null,
    val populationHigher: Int? = null,
    val specId: String? = null,
    val cardId: String? = null,
    val providerResponseHash: String,
    val retrievedAt: String,
    val refreshAfter: String,
    val cached: Boolean,
)

@Serializable
data class PsaCertIntakeRequest(
    val binderId: String,
    val entryId: String,
    val gradingCompany: String,
    val gradingScore: String? = null,
    val certNumber: String,
)

@Serializable
data class PsaIntakeResult(
    val id: String,
    val name: String,
)

@Serializable
data class PrintedIdentityUpdateRequest(
    val printedName: String? = null,
    val searchAliases: List<String> = emptyList(),
)

@Serializable
data class PrintedIdentityResult(
    val id: String,
    val name: String,
    val printedName: String? = null,
    val searchAliases: List<String> = emptyList(),
)

@Serializable
data class TrackedPriceRequest(
    val items: List<TrackedPriceItem>,
    val force: Boolean = false,
    val source: String = "automatic",
)

@Serializable data class TrackedPriceItem(val tcg: String, val externalId: String)

@Serializable
data class PriceOriginalQuote(
    val amount: Double,
    val currency: String,
    val source: String,
    val asOf: String? = null,
)

@Serializable
data class PriceFxProvenance(
    val fromCurrency: String,
    val toCurrency: String,
    val rate: Double,
    val source: String,
    val asOf: String,
)

@Serializable
data class PriceMatchProvenance(
    val method: String,
    val confidence: Double,
    val ambiguous: Boolean? = null,
    val providerProductId: String? = null,
    val providerGroupId: String? = null,
)

@Serializable
data class PriceResultProvenance(
    val provider: String,
    val retrievedAt: String,
    val originalQuotes: List<PriceOriginalQuote> = emptyList(),
    val fx: PriceFxProvenance? = null,
    val match: PriceMatchProvenance? = null,
)

@Serializable
data class TrackedPriceResult(
    val key: String,
    val tcg: String,
    val externalId: String,
    val price: Double? = null,
    val currency: String? = null,
    val source: String? = null,
    val updatedAt: String? = null,
    val cached: Boolean,
    val error: String? = null,
    val provenance: PriceResultProvenance? = null,
)

@Serializable
data class TrackedPricesEnvelope(
    val prices: List<TrackedPriceResult> = emptyList(),
    val refreshedAt: String? = null,
    val refreshAfter: String? = null,
)

object ExactCentAllocator {
    fun allocate(totalCents: Long, items: List<AcquisitionCostSplitItem>): Map<String, Long> {
        require(totalCents >= 0) { "Total must not be negative" }
        require(items.isNotEmpty()) { "At least one item is required" }
        require(items.all { it.weight > 0 }) { "Weights must be positive" }
        require(items.map { it.collectionEntryId }.distinct().size == items.size) {
            "Collection entries must be unique"
        }

        val totalWeight = items.sumOf { it.weight.toLong() }
        val base = items.associate { item ->
            item.collectionEntryId to totalCents * item.weight / totalWeight
        }.toMutableMap()
        var remainder = totalCents - base.values.sum()

        items.sortedWith(
            compareByDescending<AcquisitionCostSplitItem> {
                (totalCents * it.weight) % totalWeight
            }.thenBy { it.collectionEntryId },
        ).forEach { item ->
            if (remainder > 0) {
                base[item.collectionEntryId] = base.getValue(item.collectionEntryId) + 1
                remainder -= 1
            }
        }
        return base
    }
}

File: mobile-apps/ios/TCGer/TCGer/Models/LibraryOperationsModels.swift
import Foundation

enum StorageContainerKind: String, Codable, CaseIterable, Identifiable, Sendable {
    case binder
    case box
    case `case`
    case other

    var id: String { rawValue }
    var title: String { rawValue.capitalized }
}

struct StoragePlacement: Identifiable, Codable, Hashable, Sendable {
    let id: String
    let collectionEntryId: String
    let slotIndex: Int
    let quantity: Int
    let stackKey: String?
    var cardName: String?
    var printedName: String?
}

struct StorageCompartment: Identifiable, Codable, Hashable, Sendable {
    let id: String
    let label: String
    let order: Int
    let pageNumber: Int?
    let rows: Int
    let columns: Int
    let capacity: Int
    let locked: Bool
    let placements: [StoragePlacement]
}

struct StorageContainer: Identifiable, Codable, Hashable, Sendable {
    let id: String
    let binderId: String?
    let name: String
    let kind: StorageContainerKind
    let order: Int
    let isUnsorted: Bool
    let locked: Bool
    let compartments: [StorageCompartment]
}

struct CreateStorageContainerRequest: Encodable, Sendable {
    let name: String
    let kind: StorageContainerKind
    let binderId: String?
    let order: Int?
    let isUnsorted: Bool
    let locked: Bool
}

struct UpdateStorageContainerRequest: Encodable, Sendable {
    let name: String?
    let order: Int?
    let locked: Bool?
}

struct CreateStorageCompartmentRequest: Encodable, Sendable {
    let containerId: String
    let label: String
    let order: Int
    let pageNumber: Int?
    let rows: Int
    let columns: Int
    let capacity: Int
    let locked: Bool
}

struct UpdateStorageCompartmentRequest: Encodable, Sendable {
    let label: String?
    let order: Int?
    let pageNumber: Int?
    let locked: Bool?
}

struct PlaceCollectionEntryRequest: Encodable, Sendable {
    let compartmentId: String
    let collectionEntryId: String
    let slotIndex: Int
    let quantity: Int
    let allowDuplicateStacking: Bool
}

struct DeckCheckoutAllocation: Identifiable, Codable, Hashable, Sendable {
    let id: String
    let deckCardId: String
    let collectionEntryId: String
    let quantity: Int
    let containerId: String?
    let containerName: String?
    let compartmentId: String?
    let compartmentLabel: String?
    let slotIndex: Int?
    let cardName: String?
    let printedName: String?
    let refilledAt: String?

    var locationDescription: String {
        let location = [containerName, compartmentLabel].compactMap { $0 }.joined(separator: " · ")
        guard let slotIndex else { return location.isEmpty ? "Unsorted" : location }
        return [location, "Slot \(slotIndex + 1)"].filter { !$0.isEmpty }.joined(separator: " · ")
    }
}

struct DeckCheckoutSession: Identifiable, Codable, Hashable, Sendable {
    let id: String
    let deckId: String
    let status: String
    let note: String?
    let checkedOutAt: String
    let checkedInAt: String?
    let allocations: [DeckCheckoutAllocation]

    var isCheckedOut: Bool { status == "checked_out" }
}

struct RapidSetEntryRequest: Encodable, Sendable {
    struct Entry: Encodable, Sendable {
        let rowId: String
        let collectorNumber: String
        let card: CardData
        let quantity: Int
    }

    struct CardData: Encodable, Sendable {
        let name: String
        let printedName: String?
        let searchAliases: [String]?
        let tcg: String
        let externalId: String
        let setCode: String?
        let setName: String?
        let collectorNumber: String?
        let imageUrl: String?
        let imageUrlSmall: String?
    }

    let binderId: String
    let tcg: String
    let setCode: String
    let entries: [Entry]
}

struct RapidSetEntryReceipt: Identifiable, Codable, Hashable, Sendable {
    let receiptId: String
    let addedRows: Int
    let addedCopies: Int
    let items: [Item]
    var cardName: String?
    var printedName: String?

    var id: String { receiptId }

    struct Item: Identifiable, Codable, Hashable, Sendable {
        let rowId: String
        let collectorNumber: String
        let entryId: String
        let auditId: String
        let quantity: Int

        var id: String { auditId }
    }
}

struct UndoCollectionMutationRequest: Encodable, Sendable {
    let idempotencyKey: String
}

struct UndoCollectionMutationResponse: Codable, Sendable {
    let audit: JSONValue
}

struct LegacyRapidSetEntryReceipt: Identifiable, Codable, Hashable, Sendable {
    let id: String
    let auditId: String
    let collectionEntryIds: [String]
    let cardName: String
    let printedName: String?
    let setCode: String
    let collectorNumber: String
    let quantity: Int
}

struct AcquisitionCostSplitItem: Identifiable, Codable, Hashable, Sendable {
    let collectionEntryId: String
    var weight: Int
    var cardName: String?

    var id: String { collectionEntryId }
}

struct AcquisitionCostSplitRequest: Encodable, Sendable {
    struct Line: Encodable, Sendable {
        let collectionEntryId: String
        let weight: Int
    }

    let totalCents: Int
    let currency: String
    let mode: String
    let lines: [Line]
    let notes: String?
}

struct AcquisitionCostAllocation: Identifiable, Codable, Hashable, Sendable {
    let collectionEntryId: String
    let allocatedCents: Int
    let acquisitionPrice: Double
    let transactionId: String

    var id: String { collectionEntryId }
}

struct AcquisitionCostSplitReceipt: Identifiable, Codable, Hashable, Sendable {
    let allocationGroupId: String
    let auditId: String
    let totalCents: Int
    let currency: String
    let allocations: [AcquisitionCostAllocation]

    var id: String { allocationGroupId }
}

struct PSACertificationLookup: Codable, Hashable, Sendable {
    let certNumber: String
    let grader: String
    let grade: Double?
    let gradeLabel: String?
    let labelType: String?
    let year: String?
    let brand: String?
    let subject: String?
    let searchableName: String?
    let cardNumber: String?
    let variety: String?
    let category: String?
    let population: Int?
    let populationHigher: Int?
    let specId: String?
    let cardId: String?
    let providerResponseHash: String
    let retrievedAt: String
    let refreshAfter: String
    let cached: Bool
}

struct PSACertIntakeRequest: Encodable, Sendable {
    let binderId: String
    let entryId: String
    let gradingCompany: String
    let gradingScore: String?
    let certNumber: String
}

struct PrintedIdentityUpdateRequest: Encodable, Sendable {
    let printedName: String?
    let searchAliases: [String]
}

struct TrackedPriceRequest: Encodable, Sendable {
    struct Item: Encodable, Sendable {
        let tcg: String
        let externalId: String
    }

    let items: [Item]
    let force: Bool
    let source: String
}

struct LibraryTrackedPriceResult: Identifiable, Codable, Hashable, Sendable {
    struct Provenance: Codable, Hashable, Sendable {
        struct OriginalQuote: Codable, Hashable, Sendable {
            let amount: Double
            let currency: String
            let source: String
            let asOf: String?
        }

        struct FX: Codable, Hashable, Sendable {
            let fromCurrency: String
            let toCurrency: String
            let rate: Double
            let source: String
            let asOf: String
        }

        struct Match: Codable, Hashable, Sendable {
            let method: String
            let confidence: Double
            let ambiguous: Bool?
            let providerProductId: String?
            let providerGroupId: String?
        }

        let provider: String
        let retrievedAt: String
        let originalQuotes: [OriginalQuote]
        let fx: FX?
        let match: Match?
    }

    let key: String
    let tcg: String
    let externalId: String
    let price: Double?
    let currency: String?
    let source: String?
    let updatedAt: String?
    let cached: Bool
    let error: String?
    let provenance: Provenance?

    var id: String { key }
}

struct LibraryTrackedPricesEnvelope: Decodable, Sendable {
    let prices: [LibraryTrackedPriceResult]
    let refreshedAt: String?
    let refreshAfter: String?
}
